Dynamic pipelines v2

ranj063 opened the pull request:
This replaces #2371

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -3597,11 +3612,38 @@ int snd_sof_complete_pipeline(struct device *dev,
+			dev_err(sdev->dev, "error: volatile control found for dynamic widget %s\n",
+				comp_swidget->widget->sname);
+			return -EINVAL;
+		}
```
this could be just
```
if (scontrol->comp_id == comp_swidget->comp_id &&
    (scontrol->access & SNDRV_CTL_ELEM_ACCESS_VOLATILE)) {
	dev_err(sdev->dev, "error: volatile control found for dynamic widget %s\n",
			comp_swidget->widget->sname);
	return -EINVAL;
}
```

ranj063 replied:
done

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -3597,11 +3612,38 @@ int snd_sof_complete_pipeline(struct device *dev,
+
+	/* set the pipe_widget and apply the dynamic_pipeline_widget_flag */
+	comp_swidget->pipe_widget = pipe_widget;
+	comp_swidget->dynamic_pipeline_widget = pipe_widget->dynamic_pipeline_widget;
```
this looks safe now that you made `.complete` return an error code and you have the caller check it, but still it might be more robust to first do the checks and only then begin modifying persistent data, like
```
if (pipe_widget->dynamic_pipeline_widget) {
	/* dynamic widgets cannot have volatile kcontrols */
	list_for_each_entry(scontrol, &sdev->kcontrol_list, list)
		...
}

/* set the pipe_widget and apply the dynamic_pipeline_widget_flag */
comp_swidget->pipe_widget = pipe_widget;
comp_swidget->dynamic_pipeline_widget = pipe_widget->dynamic_pipeline_widget;
```

ranj063 replied:
done

lyakh reviewed (commented):
just a couple of minor comments

kv2019i commented on include/sound/soc-topology.h:
```diff
@@ -151,7 +151,7 @@ struct snd_soc_tplg_ops {
 
 	/* completion - called at completion of firmware loading */
-	void (*complete)(struct snd_soc_component *);
+	int (*complete)(struct snd_soc_component *comp);
```
Re-reading all the commits again. It's not so obvious how the return value is handled after the change. It's somewhat non-obvious until you look at the changes in soc_tplg_load(), and check that in fact all callers from fw load do pass errors up. I.e. with this patch, error will propagate all the way up.

Maybe add a sentence that you are also modifying soc-topology.c to ensure the return value is passed all the way up.

kv2019i commented on include/uapi/sound/sof/tokens.h:
```diff
@@ -51,6 +51,7 @@
 #define SOF_TKN_SCHED_CORE			203
 #define SOF_TKN_SCHED_FRAMES			204
 #define SOF_TKN_SCHED_TIME_DOMAIN		205
+#define SOF_TKN_SCHED_DYNAMIC_PIPELINE		206
```
Hmm, I was about to continue complaining about the name. Ondemaind-versus-preloaded would seem more appropriate names. But, but, in software libraries, concept of linking dynamically versus statically is well established. So maybe this is good after all. It's just that in context of DSP pipelines, "dynamic" could have many other meanings.

kv2019i reviewed (commented):
Thanks @ranj063 , lot of good improvements. I'll still need more time for the last two patches. But e.g. the patch to ensure DAI_CONFIG is not sent before widget is setup, is much easier to follow. The last patch I already reviewed before, but I can go through it again.

lgirdwood commented:
@ranj063 just to let you know that @RanderWang has a patch opn top of this for IPC flow alignment that moves PIPE_NEW before the components are added. Will align stuff better with other IPC ABIs.

RanderWang commented:
> @ranj063 just to let you know that @RanderWang has a patch opn top of this for IPC flow alignment that moves PIPE_NEW before the components are added. Will align stuff better with other IPC ABIs.

@ranj063 I just created pipeline first for SOF FW in widget_list_setup and it works. It enables SOF FW go with the same sequnce of CAVS

kv2019i commented on sound/soc/sof/sof-audio.c:
```diff
@@ -204,6 +229,244 @@ int sof_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
+
+		/*
+		 * free widget and its pipe_widget. Either of these can fail, but free as many as
+		 * possible before freeing the list and returning the error.
```
Use of "free" in the SOF FW IPC sense, is a bit confusing here again. From a kernel point of view, trying to free as many as possible, sounds very, very scary. But as here "free" relates to sending the free ipc, I think this is very reasonable. The naming is what it is and you have the comment there, so I'm good with this.

kv2019i reviewed (dismissed):
Thanks @ranj063 , looks good now! 

If we are all good, let's follow up with merging https://github.com/thesofproject/sof/pull/3905 to ensure CI coverage.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -87,8 +89,19 @@ struct snd_sof_widget {
+	/*
+	 * Flag indicating if the widget should be set up dynamically when a PCM is opened.
+	 * This flag is only set for the scheduler type widget in topology. During topology
+	 * loading, this flag is propagated to all the widgets belonging to tthe same pipeline.
```
typo: tthe

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -3597,11 +3612,36 @@ int snd_sof_complete_pipeline(struct device *dev,
 	return 1;
 }
 
+/* set pipe_widget for comp */
```
it took me a while to figure out what 'comp' was. I understood the comment as 'set pipe_widget for comparison'.
Avoid abbreviations in comments...

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -3597,11 +3612,36 @@ int snd_sof_complete_pipeline(struct device *dev,
 }
 
+/* set pipe_widget for comp */
+static int sof_set_comp_pipe_widget(struct snd_sof_dev *sdev, struct snd_sof_widget *pipe_widget,
```
set_comp_pipe_widget or set_pipe_comp_widget?

you have 3 terms that could be used in different ways, what is the precedence order?

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -216,6 +218,7 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
 	if (runtime_suspend)
 		sof_cache_debugfs(sdev);
 #endif
+
```
spurious line change?

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -316,6 +317,16 @@ int sof_restore_pipelines(struct device *dev)
+	struct snd_sof_route *sroute;
+
+	list_for_each_entry_reverse(sroute, &sdev->route_list, list)
+		sroute->setup = false;
```
is any lock needed to protect the list? add a comment if not.

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -157,7 +157,7 @@ static int sof_resume(struct device *dev, bool runtime_resume)
 
 	/* restore pipelines */
-	ret = sof_restore_pipelines(sdev->dev);
+	ret = sof_set_up_pipelines(sdev->dev);
```
typos in commit message: pipelies

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -324,12 +382,19 @@ int sof_set_up_pipelines(struct device *dev)
 	struct snd_sof_route *sroute;
 
+	list_for_each_entry_reverse(swidget, &sdev->widget_list, list)
+		swidget->use_count = 0;
```
does the 'reverse' matter if you are not doing any operation before the entire list is checked?

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -316,6 +317,16 @@ int sof_restore_pipelines(struct device *dev)
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+	struct snd_sof_route *sroute;
+
+	list_for_each_entry_reverse(sroute, &sdev->route_list, list)
```
does the reverse matter and why?

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -672,9 +672,31 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_control *scontrol,
 	size_t send_bytes;
 	int err;
 
+	list_for_each_entry_reverse(swidget, &sdev->widget_list, list) {
```
again why reverse?

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -153,47 +153,41 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
 /* Send DAI_CONFIG IPC to the DAI that matches the dai_name and direction */
 static int hda_link_config_ipc(struct sof_intel_hda_stream *hda_stream,
-			       const char *dai_name, int channel, int dir)
+			       const char *dai_name, int channel, int dir, bool setup)
```
You are repeating yourself in the commit message, the first and second sentence say the same thing.

Conversely the second paragraph is quite unclear.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -53,67 +53,107 @@ static int sdw_clock_stop_quirks = SDW_INTEL_CLK_STOP_BUS_RESET;
+	config->dai_index = (link_id << 8) | d->id;
+	config->alh.stream_id = alh_stream_id;
+
+	return hda_dai_widget_update(sdev, sof_dai, config, setup, dir);
```
the naming doesn't seem too good, why would SoundWire have anything to do with hda? the prefix looks odd and wrong at the same time.

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -259,6 +300,13 @@ static int sof_pcm_hw_free(struct snd_soc_component *component,
 	}
 
+	/*
+	 * for static pipelines, the widgets won't be freed and only the DAI_CONFIG will be updated
```
updated to what?

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -414,8 +464,15 @@ static int sof_pcm_trigger(struct snd_soc_component *component,
+		if (ret < 0)
+			return ret;
+
+		/* free widget list during suspend */
```
I don't understand the comment and what suspend has to do with trigger?

ranj063 replied:
sorry that was incorrect. I meant to say SUSPEND trigger not suspend.

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -82,6 +82,15 @@ static int sof_widget_kcontrol_setup(struct snd_sof_dev *sdev, struct snd_sof_wi
+
+	list_for_each_entry(sroute, &sdev->route_list, list)
+		if (sroute->src_widget == widget || sroute->sink_widget == widget)
+			sroute->setup = false;
```
i am pretty sure I've seen this before, is this a bad squash?

ranj063 replied:
No, this is legit. What you saw before was applicable only for static pipelines. This is where the route status is updated for dynamic pipelines.

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -97,8 +106,8 @@ int sof_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
-	/* only free when refcount is 0 */
-	if (--swidget->use_count)
+	/* only free when use_count is 0 */
+	if (--swidget->use_count > 0)
```
this looks like it should be squashed earlier, and I also don't get what the > 0 brings?

plbossart reviewed (changes requested):
lots of minor comments and nit-picks below.
I am not sure if it's a GitHub issue but I still don't have a good sense of which patches are preparatory and don't change functionality, and which ones do.
I am also confused about transitions in general, e.g. what happens on a resume while playback is on-going or resume while paused.

One more round and we should be good to go.

RanderWang commented on sound/soc/sof/topology.c:
```diff
@@ -1585,7 +1581,6 @@ static int sof_widget_load_buffer(struct snd_soc_component *scomp, int index,
 				  struct snd_soc_tplg_dapm_widget *tw,
 				  struct sof_ipc_comp_reply *r)
```
r  is not used now, so we can remove it

RanderWang replied:
And also you need to modify the code referring this r.

ranj063 commented:
> lots of minor comments and nit-picks below.
> I am not sure if it's a GitHub issue but I still don't have a good sense of which patches are preparatory and don't change functionality, and which ones do.
> I am also confused about transitions in general, e.g. what happens on a resume while playback is on-going or resume while paused.
> 
> One more round and we should be good to go.

@plbossart sorry for the delay. I've updated the PR addressing all your comments now.

lyakh commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -153,47 +153,48 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
 
-	return -EINVAL;
+	/* set/free up DAI widget and send DAI_CONFIG IPC */
+	return hda_sdw_dai_widget_setup(sdev, sof_dai, config, widget_setup, dir);
```
is it just me who finds it unclear why we here are suddenly dealing with soundwire? So far the code seems to be generic HDA, why is it now SDW-specific? Maybe a comment could help...

ranj063 replied:
this patch was always there just with a different naming?

lyakh replied:
that's possible that I failed to comment on that in the previous versions, sorry about that. And thanks for changing!

lyakh commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -296,7 +297,6 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
 	link = snd_hdac_ext_bus_get_link(bus, asoc_rtd_to_codec(rtd, 0)->component->name);
 	if (!link)
 		return -EINVAL;
-
```
left-over?

lyakh commented on sound/soc/sof/intel/hda.c:
```diff
@@ -53,67 +53,114 @@ static int sdw_clock_stop_quirks = SDW_INTEL_CLK_STOP_BUS_RESET;
+		if (!sof_dai->cpu_dai_name)
+			continue;
+
+		if (!strcmp(d->name, sof_dai->cpu_dai_name)) {
```
nitpick: could be
```
if (sof_dai->cpu_dai_name && !strcmp(d->name, sof_dai->cpu_dai_name)) {
	found = true;
	break;
}
```

lyakh commented on sound/soc/sof/intel/hda.c:
```diff
@@ -53,67 +53,114 @@ static int sdw_clock_stop_quirks = SDW_INTEL_CLK_STOP_BUS_RESET;
+		return ret;
+
+	/* and then free the widget */
+	return sof_widget_free(sdev, swidget);
```
This is intended, right? In a `*_setup()` method in success case the last action is freeing the widget, and the return status of that operation is our return code. If that's indeed intended and correct, I think a comment is needed here.

lyakh replied:
Ok, after some more reading I understand, that some "*_setup()" functions can now do either set up / initialisation or freeing depending on their `setup` parameter. But `sof_widget_setup()` is only for setup / init. Seems a bit confusing to me... Maybe the naming could be improved, but no, I have no good ideas, sorry. Maybe something along the lines of configure / reconfigure / update / amend...

ranj063 replied:
I've added a comment for the function to call this out but left the name unchanged.

lyakh commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -153,47 +153,48 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
+					   &reply, sizeof(reply));
 
-	return -EINVAL;
+	/* set/free up DAI widget and send DAI_CONFIG IPC */
```
"set up / free" maybe?

lyakh commented on sound/soc/sof/topology.c:
```diff
@@ -2479,47 +2381,40 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
-		/* Find scontrol for this pga and set readback offset*/
-		list_for_each_entry(scontrol, &sdev->kcontrol_list, list) {
-			if (scontrol->comp_id == swidget->comp_id) {
-				scontrol->readback_offset = reply.offset;
```
where is `.readback_offset` now set?

ranj063 replied:
not anywhere but it was never used before and I never understood what it was for even before.

lyakh replied:
but it is used in ipc.c to "write/read value header via mmaped region"

kv2019i replied:
Yup, @lyakh and @ranj063, this is used in snd_sof_ipc_set_get_comp_data()

ranj063 commented:
@juimonen this PR is now updated to also account for SSP BE hw_params/free.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -237,8 +237,8 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_control *scontrol,
 /* PM */
-int sof_restore_pipelines(struct device *dev);
 void sof_tear_down_pipelines(struct device *dev);
+int sof_set_up_pipelines(struct device *dev);
```
move two lines above to show it's just a rename.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -237,8 +237,8 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_control *scontrol,
 /* PM */
-int sof_restore_pipelines(struct device *dev);
 void sof_tear_down_pipelines(struct device *dev);
+int sof_set_up_pipelines(struct device *dev);
```
also wondering if we should use startup/shutdown which seems to be the usual language in ASoC for this kind of stuff.

ranj063 replied:
hmm Im not sure if we use startup/shutdown then we'd have to move this to the startup/shutdown callbacks and Im not sure if that makes sense

plbossart replied:
ok, we are not trying to compete for the next inauguration poem :-)

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -324,16 +382,23 @@ int sof_set_up_pipelines(struct device *dev)
-	 * No need to protect this function is called only during the suspend callback and all
-	 * streams should be suspended by then
+	 * No need to protect swidget->use_count and sroute->setup as this function is called only
+	 * during the suspend callback and all streams should be suspended by then
```
not sure what the initial comment was supposed to mean, maybe update the initial patch to be clearer.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -153,47 +153,48 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
 
-	return -EINVAL;
+	/* set up/free DAI widget and send DAI_CONFIG IPC */
+	return hda_sdw_dai_widget_setup(sdev, sof_dai, config, widget_setup, dir);
```
I also don't get what this -sdw- thing does here. Is this intentional?

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -153,47 +153,48 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
 
-	return -EINVAL;
+	/* set up/free DAI widget and send DAI_CONFIG IPC */
+	return hda_sdw_dai_widget_setup(sdev, sof_dai, config, widget_setup, dir);
```
the commit message says "dai widget setup before IPC'. I read the exact opposite with IPC before setup in lines 191..197.

Confused.

ranj063 replied:
line 193 does only DAI_CONFIG IPC without anything to do with the DAI widget. This is needed to send the invalid config in the stop trigger. The widget is set up floowed by DAI_CONFIG IP during hw_params and is freed in hw_free after sending the invalid DAI_CONFIG IPC.

plbossart replied:
sorry @ranj063 this is clear as mud. please consider listing out the steps that are done today and what needs to happens after the change to dynamic pipelines. You are also not describing what 'config_ipc_only' might possibly refer to. there's only a true/false setting for that flag, it's not self-explanatory at all. You may want to add a comment for each true/false argument to make the reviews easier for the rest of us.

also you have a typo in the commit message s/BW/BE hw_params?

ranj063 replied:
ok ive redone this part. please let me know if it looks clearer.

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -259,6 +300,11 @@ static int sof_pcm_hw_free(struct snd_soc_component *component,
 	}
 
+	/* for static pipelines, the widgets won't be freed as the use_count will never be 0 */
+	ret = sof_widget_list_free(sdev, spcm, substream->stream);
```
I can't relate the comment with the function, or you are describing what happens inside of the function?

ranj063 replied:
I've removed the comment. Maybe its better without it.

plbossart replied:
indeed.

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -653,24 +653,37 @@ int sof_set_up_pipelines(struct device *dev)
+ * routes and use_count for all widgets.
  */
-void sof_tear_down_pipelines(struct device *dev)
+int sof_tear_down_pipelines(struct device *dev, bool verify)
```
Sorry, I am missing the concept on why we need a verify flag on teardown?

It's useful on setup to load all pipelines and check what the topology does. But on suspend I don't get what you are trying to do.

ranj063 replied:
The flag only tells you whether you need to tear down the pipelines or not. Previously we have only one case where we called tear_down_pipelines and that was during suspend. And during suspend we dont free anything, we just simply power down the DSP.

But now that we want to do a tplg verification during load, we should tear it down after the verification is complete. And when we tear the pipelines down during verification, we should actually free the widgets and so the verify flag needs to used for both set up and teardown

plbossart replied:
ah ok, so you are also verifying the IPC free, but do we expect any errors, or is it just to go back to a blank slate?

ranj063 replied:
Yes, we do expect errors during verification and we return the error so the card wont be registered.

plbossart reviewed (commented):
Minor comments @ranj063 

No fundamental objections, just clarifications for consistency/clarity

lyakh commented:
The `.readback_offset` issue is the last question to clarify for me in this PR

kv2019i commented on sound/soc/sof/intel/hda.c:
```diff
@@ -41,6 +41,62 @@
+	/*
+	 * For static pipelines, the DAI widget would already be set up and calling
+	 * sof_widget_setup()/free() would simply return without doing anything.
+	 * For dynamic pipelines, the DAI widget would be set up or freed here.
```
The conditional "would be set up here" is a bit weird. I mean this reads as a placeholder for the actual code. The actions are already implemented, so "For dynamic pipelines, the DAI widget is set up or freed here." would be more accurate, right.

kv2019i reviewed (changes requested):
With 12 commits and a lot of code touched, this is getting pretty hard to review.

I did not find anything new in latest review round (one minor comment on code comment, but no need to update the series for this).

@lyakh Did find a good point on "scontrol->readback_offset" -> this needs to be checked before merge.

So except that one thing, I'm ok to approve the series.

ranj063 commented:
> The `.readback_offset` issue is the last question to clarify for me in this PR

@lyakh @kv2019i the readback off set is "used" as in it does something when it is non-zero in ipc.c. But we've never set it to anything other than 0 AFAIK. In fact the other flaw we've had is that we set up the PGA widget and set the readback_offset during topology loading but never did it after resuming from D3. But as far as this commit goes, it is the right thing because the PGA widget is not really set up in the FW so you cant set the readback offset.

kv2019i commented on sound/soc/sof/topology.c:
```diff
@@ -2286,32 +2227,6 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
-
-	/* we sent the data in single message so return */
-	if (ipc_data_size)
-		goto out;
```
Hmm, while trying to understand readback_offset, I noticed this is removing some special logic in case ipc_data_size is larger than IPC message. In the code you remove, if there's more data, ipc_data_size is set to zero in above code L2245 and we send an IPC message out. 

Here it seems we store the data only if it fits a single IPC message. This seems like the only place where we call sof_get_control_data(), so it seems this init control data is lost.

ranj063 replied:
ahh good catch. let me fix this.

ranj063 replied:
fixed now. process->size should be set to ipc_size always

ranj063 replied:
@kv2019i sorry I spoke too soon. The original setting was correct. process->size = ipc_data_size when the total size is less than IPC_MAX. otherwise it should be 0

kv2019i replied:
@ranj063 Hmm, but where is this init data passed now if it doesn't with the IPC msg?

ranj063 replied:
If the control data size is larger than max ipc size, it will be sent when widget controls are set up when the pcm is started. If not, it's copied as part of the process_upc_comp and the ipc_data_dize is set here accordingly

kv2019i reviewed (changes requested):
@ranj063 wrote:
> @lyakh @kv2019i the readback off set is "used" as in it does something when it is non-zero in ipc.c. But we've never set it to anything other than 0 AFAIK. In fact the other flaw we've had is that we set up the PGA widget and set the readback_offset during topology loading but never did it after resuming from D3. But as far as this commit goes, it is the right thing because the PGA widget is not really set up in the FW so you cant set the readback offset.

Hmm, checked fw master and indeed, the comp_reply.offset is never set. So this seems completely stale code.

I'd prefer to not tackle this in this series. It's not easy as you are essentially removing the only occasion where "reply.offset" is used in this patch. But as you note, it's broken anyways. Hmm, ok with this patch.

While browsing the code for readback_offset, I did find one new potential issue, please see inline.

lyakh commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -151,49 +151,77 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
+
+	sof_dai = hda_get_sof_dai(hda_stream, dai_name, channel, dir);
+	if (IS_ERR(sof_dai))
+		return PTR_ERR(sof_dai);
```
You never return `ERR_PTR()` from `hda_get_sof_dai()` you return `NULL` on error.

ranj063 replied:
yes, thanks. I can never get the ERR_PTR()/PTR_ERR() bits right!

lyakh commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -151,49 +151,77 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
+
+	sof_dai = hda_get_sof_dai(hda_stream, dai_name, channel, dir);
+	if (IS_ERR(sof_dai))
+		return PTR_ERR(sof_dai);
```
ditto

kv2019i reviewed (approved):
As kcontrol restore works after D3->D0 state change, removing the code in widget_ready (which is only called at tplg load), seems fine after all. 

With that, I'm fine with the series.

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -672,9 +672,31 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_control *scontrol,
+	 */
+	if (!swidget->use_count)
+		return 0;
+
```
Out of curiosity, in which cases would we have widgets that belong to multiple pipelines?

I can only think of bi-directional DAIs, are there other examples?

ranj063 replied:
even mixers/mux. basically the list of connected DAPm widgets for 2 PCMs that share a mixer will both contain the mixer. So if one PCM is started, the second one should skip setting it up and oncversely while freeing, if one is using it, we should skip freeing it

plbossart replied:
Mux and mixers will have their sources in a set of pipelines but their sink is in a different one. Does this make the mux and mixers "belong" to different pipelines?  What does "belong" mean really?

ranj063 replied:
belong is probably misleading yout into thinking that the widget is part of both pipelines in topology. WHat I really mean is a widget like the mixer will "belong" to the list of of connected DAPM widgets for both the PCMs that use the mixer widget.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -53,67 +109,58 @@ static int sdw_clock_stop_quirks = SDW_INTEL_CLK_STOP_BUS_RESET;
+	config->dai_index = (link_id << 8) | d->id;
+	config->alh.stream_id = alh_stream_id;
+
+	return hda_dai_widget_setup(sdev, sof_dai, config, setup, dir);
```
we've got to find a better naming here.

It's not intuitive at all you use an 'hda_dai_widget' function to configure a SoundWire DAI which uses a different DMA and different ALH IP.

ranj063 replied:
if I move it to sof-audio.c and call is just sof_dai_widget_update(), would that be OK?

plbossart replied:
Ah, I get the problem now. hda can be a link or can be a controller (handled in hda.c).
I think you are using "hda" as a namespace prefix here for the controller, it has nothing to do with hdaudio link/dai.

rename it as hda_ctlr_dai_widget_setup?

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -204,6 +229,244 @@ int sof_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
+
+		ret = sof_widget_free(sdev, swidget->pipe_widget);
+		if (ret < 0)
+			ret1 = ret;
```
I don't fully get the use of ret1, what are you trying to achieve? This will return the last error, but did we want to return the first?

lyakh replied:
I think what @ranj063 is trying to achieve here is not losing an error. Say, if one of these `sof_widget_free()` calls does return an error, we still want to continue freeing other widgets, then the next one can return 0, but we still want to return an error to indicate, that something didn't quite work out. As for whether to return the first or the last one - in principle I agree, that the first one would probably be more meaningful, but in practice this probably won't matter much, hopefully we won't get too many errors here.

plbossart reviewed (dismissed):
I was inches away from approving this PR @ranj063 but I still had a couple of nit-picks. Nothing fundamental really, the flow looks good to me overall.

juimonen reviewed (commented):
nitpick: in "ASoC: SOF: Don't set up widgets during topology parsing" check the commit message, there's some extra spaces and maybe some extra commas can be removed

juimonen reviewed (commented):
nitpick: in "ASoC: SOF: Add new fields to snd_sof_route", It feels that commit message is somewhat "diminishing" what is actually done -> so maybe somehow make it say in the title that also new function is added etc. I see this is coming from maybe chopping the PR into separate commits...

plbossart reviewed (approved):
I think it's good enough for merges and tests now @ranj063 

We may have to stress test a bit more with HDaudio and SoundWire topologies before sending upstream in the next cycle

ranj063 commented:
> I think it's good enough for merges and tests now @ranj063
> 
> We may have to stress test a bit more with HDaudio and SoundWire topologies before sending upstream in the next cycle

@plbossart sure, we can start switching one platform at a time to use dynamic pipes in topology. I already have a PR pending for HDA topologies to switch to dynamic. Once this is merged, we can start testing.

kv2019i reviewed (approved):
@ranj063 Do you want to fix the typos Jaska noted or good to go with this one? I think those were minor enough, so feel free to merge if you think this is ready.

Also, can you resubmit the SOF patch to enable dynamic loading for HDA topologies. It got closed due to the branch renaming in sof.git.

plbossart commented:
ok, we've reviewed this enough, let's test it for a couple of weeks with HDAudio + SoundWire platforms. Baytrail would be interesting as well since we are not using pm_runtime so don't have the safety net that comes with periodic firmware reboots w/ pm_runtime suspend/resume.
